packages/backend/src/open_payments/grantReference/service.test.ts

```diff
@@ -0,0 +1,84 @@
+import { Knex } from 'knex'
+import { v4 as uuid } from 'uuid'
+
+import { IocContract } from '@adonisjs/fold'
+import { createTestApp, TestContainer } from '../../tests/app'
+import { initIocContainer } from '../..'
+import { AppServices } from '../../app'
+import { Config } from '../../config/app'
+import { GrantReferenceService } from './service'
+import { truncateTables } from '../../tests/tableManager'
+import { GrantReference } from './model'
+
+describe('Grant Reference Service', (): void => {
+  let deps: IocContract<AppServices>
+  let appContainer: TestContainer
+  let grantReferenceService: GrantReferenceService
+  let knex: Knex
+  let knexTrx: Knex.Transaction
+
+  beforeAll(async (): Promise<void> => {
+    deps = await initIocContainer(Config)
+    appContainer = await createTestApp(deps)
+    knex = await deps.use('knex')
+  })
+
+  beforeEach(async (): Promise<void> => {
+    grantReferenceService = await deps.use('grantReferenceService')
+  })
+
+  afterEach(async (): Promise<void> => {
+    jest.useRealTimers()
+    await truncateTables(knex)
+  })
+
+  afterAll(async (): Promise<void> => {
+    await appContainer.shutdown()
+  })
+
+  describe('Create and Get GrantReference', (): void => {
+    test.each`
+      transaction
+      ${false}
+      ${true}
+    `(
+      'a grant reference can be created and fetched - transaction: $transaction',
+      async ({ transaction }): Promise<void> => {
+        const trx = transaction ? knexTrx : undefined
+        const id = uuid()
+        const grantRef = await grantReferenceService.create(
+          {
+            id,
+            clientId: uuid()
+          },
+          trx
+        )
+        const retrievedRef = await grantReferenceService.get(id, trx)
+        expect(grantRef).toEqual(retrievedRef)
+      }
+    )
+    test('cannot fetch non-existing grant reference', async (): Promise<void> => {
+      await expect(grantReferenceService.get(uuid())).resolves.toBeUndefined()
+    })
+  })
+
+  describe('Lock GrantReference', (): void => {
+    test('a grant reference can be locked', async (): Promise<void> => {
+      const grantRef = await grantReferenceService.create({
+        id: uuid(),
+        clientId: uuid()
+      })
+      const lock = async (): Promise<void> => {
+        return await GrantReference.transaction(knex, async (trx) => {
+          await grantReferenceService.lock(grantRef.id, trx)
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          await new Promise((f: any) => setTimeout(f, 6000))
+          await grantReferenceService.get(grantRef.id, trx)
```

Comment:
You could also just lock it and then do
```
await expect(grantReferenceService.get(grantRef.id, trx).skipLocked()).resolves.toBeUndefined()
```

Reply:
That is probably the faster version. 👍

Reply:
So I can't do `skipedLocked` on the `get` method, instead I put 
```
await GrantReference.transaction(async (trx) => {
  await grantReferenceService.lock(grantRef.id, trx)
   await expect(
    GrantReference.query(trx).findById(grantRef.id).skipLocked()
  ).resolves.toBeUndefined()
})
```
However, that resulted in `Rejected to value: [Error: .skipLocked() can only be used after a call to .forShare() or .forUpdate()!]`.

Reply:
You could add one of those to query or just leave the test as is.

Reply:
That also didn't work. It queries it successfully. It probably has to do with the fact that we can't do `forNoKeyUpdate()` using objection.  I'll just leave it as is.